services/ui_backend_service/data/cache/get_artifacts_action.py

```diff
@@ -100,41 +100,44 @@ def stream_error(err, id, traceback=None):
-                                results[artifact_key] = json.dumps([True, str(content)])
-                            except Exception as ex:
-                                # Exceptions might be fixable with configuration changes or other measures,
-                                # therefore we do not want to write anything to the cache for these artifacts.
-                                stream_error(str(ex), "artifact-handle-failed", get_traceback_str())
+                            # TODO: Figure out a way to store the artifact content without decoding?
+                            # presumed that cache_data/tmp/ does not persist as long as the cached items themselves,
+                            # so we can not rely on the file existing if we only return a filepath as a cached response
+                            content = decode(artifact_data.path)
+                            results[artifact_key] = json.dumps([True, content])
                         else:
                             # TODO: does this case need to raise an error as well? As is, the fetch simply fails silently.
-                            results[artifact_key] = json.dumps([False, 'object is too large'])
-                except MetaflowS3AccessDenied as ex:
-                    stream_error(str(ex), "s3-access-denied")
-                except MetaflowS3NotFound as ex:
-                    stream_error(str(ex), "s3-not-found")
-                except MetaflowS3URLException as ex:
-                    stream_error(str(ex), "s3-bad-url")
-                except MetaflowS3CredentialsMissing as ex:
-                    stream_error(str(ex), "s3-missing-credentials")
-                except MetaflowS3Exception as ex:
-                    stream_error(str(ex), "s3-generic-error", get_traceback_str())
+                            results[artifact_key] = json.dumps([False, 'artifact-too-large', artifact_data.size])
+
+                    except TypeError:
+                        # In case the artifact was of a type that can not be json serialized,
+                        # we try casting it to a string first.
+                        results[artifact_key] = json.dumps([True, str(content)])
+                    except MetaflowS3AccessDenied as ex:
+                        results[artifact_key] = json.dumps([False, 's3-access-denied', location])
+                    except MetaflowS3NotFound as ex:
+                        results[artifact_key] = json.dumps([False, 's3-not-found', location])
+                    except MetaflowS3URLException as ex:
+                        results[artifact_key] = json.dumps([False, 's3-bad-url', location])
+                    except MetaflowS3CredentialsMissing as ex:
+                        results[artifact_key] = json.dumps([False, 's3-missing-credentials', location])
+                    except MetaflowS3Exception as ex:
+                        results[artifact_key] = json.dumps([False, 's3-generic-error', get_traceback_str()])
```

Comment:
Caching all the error cases as well is very much undesirable in the long run as I think there is currently no way to invalidate a cached result?
I see no harm for this in the short term though, as the cache storage is transient due to it being part of the deployment, and if an admin is fixing S3 access issues on a deployed service, they will most likely redeploy, thus flushing the cache.